### PR TITLE
Add CI to spot_ros2 repo

### PR DESCRIPTION
**Pull request by pangtao-BDAI**
This PR adds the workflow that builds the docker image in which tests are run. The test workflow will be added in a separate PR.

### CODE REVIEW

File: .github/workflows/build_jammy_humble_essentials.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+
+      # Build and push Docker image with Buildx (don't push on PR)
+      # https://github.com/docker/build-push-action
+      - name: Build and push Docker image
+        id: build-and-push
+        uses: docker/build-push-action@ac9327eae2b366085ac7f6a2d02df8aa8ead720a
+        with:
+          context: ./docker/ros
```
Comment (jiuguangw) [inline .github/workflows/build_jammy_humble_essentials.yml:68]: This part needs to change to the correct path and file (.devcontainer/Dockerfile_essentials)

File: .github/workflows/build_jammy_humble_essentials.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
+name: Docker - Build Jammy/Humble Essentials
+
+on:
+  pull_request:
+    branches: [ "main" ]
```
Comment (jiuguangw) [inline .github/workflows/build_jammy_humble_essentials.yml:5]: Probably get rid of the `branches: [ "main" ]` (as this currently implies anything that merges into main will encounter this workflow)

File: .github/workflows/build_jammy_humble_essentials.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    paths:
+      - ".github/workflows/**"
```
Comment (jiuguangw) [inline .github/workflows/build_jammy_humble_essentials.yml:7]: Here, I would add a second line `"./devcontainer/**"` - meaning any PR that touches files within `devcontainer` will trigger this workflow.

Comment (pangtao-BDAI) [reply .github/workflows/build_jammy_humble_essentials.yml]: Done.

**Comment by jiuguangw** [comment]
Super fast review before my next meeting:

- [ ] Rename the Dockerfile to just "Dockerfile" and update the name in the workflow file.
- [ ] Rename the workflow "build_docker.yml"
- [ ] Remove a lot of the other unnecessary packages within the Dockerfile (this may involve some trial and error)

**Comment by jiuguangw** [comment]
Also,

- [ ] Change the PR title to "Aadd CI to spot_ros2 repo" (no need to mention DMM-40 since this is an external repo). For internal repos, the correct formatting is "[DMM-1234] Add feature x"

File: .github/workflows/build_jammy_humble_essentials.yml (hunk 2 above)
Comment (jiuguangw) [reply .github/workflows/build_jammy_humble_essentials.yml:5]: So my concern is [this](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions): 

> If you specify multiple activity types, only one of those event activity types needs to occur to trigger your workflow.

My interpretation of the relationship between your two filters, "branches" / "paths", is OR, not AND. If someone PRs a change to `main` but didn't touch the Dockerfile, then currently this workflow would trigger (and it shouldn't)

Comment (pangtao-BDAI) [reply .github/workflows/build_jammy_humble_essentials.yml:5]: That makes sense! `branches: [ "main" ]` has been removed.

File: .github/workflows/build_docker.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      # with sigstore/fulcio when running outside of PRs.
+      id-token: write
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      # Workaround: https://github.com/docker/build-push-action/issues/461
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@79abd3f86f79a9d68a23c75a09a9a85889262adf
```
Comment (pangtao-BDAI) [inline .github/workflows/build_docker.yml:30]: Is it the standard practice to use actions versioned by their sha, as opposed by their version numbers?

Comment (jiuguangw) [reply .github/workflows/build_docker.yml:30]: I have no idea. I've seen both.

File: devcontainer/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,51 @@
+FROM ros:humble-ros-base-jammy
```
Comment (jiuguangw) [inline devcontainer/Dockerfile:1]: Looks good but please rename the directory from `devcontainer` to `.devcontainer`. It probably doesn't matter (I've seen this named `docker` in other repos) but GitHub says [here](https://docs.github.com/en/codespaces/setting-up-your-project-for-codespaces/adding-a-dev-container-configuration/introduction-to-dev-containers):

> The configuration files for a dev container are contained in a .devcontainer directory in your repository. 

We may eventually need to enable GitHub codespaces for this repo too, so it's good to address this early.

Comment (jiuguangw) [reply devcontainer/Dockerfile:1]: Also, after the rename, you might need to change the reference in `build_docker.yml`

File: .github/workflows/build_docker.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
+name: Docker - Build Jammy/Humble Essentials
```
Comment (jiuguangw) [inline .github/workflows/build_docker.yml:1]: name: Build and publish Jammy/Humble Docker image

File: .github/workflows/build_docker.yml [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+on:
+  pull_request:
+    paths:
+      - "devcontainer/**"
```
Comment (jiuguangw) [inline .github/workflows/build_docker.yml:6]: Change this, see other comment about renaming.

File: .github/workflows/build_docker.yml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+
+      # Build and push Docker image with Buildx (don't push on PR)
+      # https://github.com/docker/build-push-action
+      - name: Build and push Docker image
+        id: build-and-push
+        uses: docker/build-push-action@ac9327eae2b366085ac7f6a2d02df8aa8ead720a
+        with:
+          context: .
+          file: devcontainer/Dockerfile
```
Comment (jiuguangw) [inline .github/workflows/build_docker.yml:57]: Change this, see other comment about renaming.

**Review by jiuguangw — COMMENTED** [review]
Overall looks great, just one minor change requested.

File: devcontainer/Dockerfile (hunk 5 above)
Comment (pangtao-BDAI) [reply devcontainer/Dockerfile:1]: Done.

**Review by jiuguangw — APPROVED** [review]
Looks good, thanks! Please make sure that after merging, it will successfully push the image to GCR (sometimes there's weird GHA permission issues and since pushing is not done during the PR, it's unclear right now)

**Pull request merged**